client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -277,60 +278,64 @@ private ResponseOrResponseException convertResponse(InternalRequest request, Nod
      * @param responseListener the {@link ResponseListener} to notify when the
      *      request is completed or fails
      */
-    public void performRequestAsync(Request request, ResponseListener responseListener) {
+    public Cancellable performRequestAsync(Request request, ResponseListener responseListener) {
         try {
             FailureTrackingResponseListener failureTrackingResponseListener = new FailureTrackingResponseListener(responseListener);
             InternalRequest internalRequest = new InternalRequest(request);
             performRequestAsync(nextNodes(), internalRequest, failureTrackingResponseListener);
+            return internalRequest.cancellable;
         } catch (Exception e) {
             responseListener.onFailure(e);
+            return Cancellable.NO_OP;
         }
     }
 
     private void performRequestAsync(final NodeTuple<Iterator<Node>> nodeTuple,
                                      final InternalRequest request,
                                      final FailureTrackingResponseListener listener) {
-        final RequestContext context = request.createContextForNextAttempt(nodeTuple.nodes.next(), nodeTuple.authCache);
-        client.execute(context.requestProducer, context.asyncResponseConsumer, context.context, new FutureCallback<HttpResponse>() {
-            @Override
-            public void completed(HttpResponse httpResponse) {
-                try {
-                    ResponseOrResponseException responseOrResponseException = convertResponse(request, context.node, httpResponse);
-                    if (responseOrResponseException.responseException == null) {
-                        listener.onSuccess(responseOrResponseException.response);
-                    } else {
-                        if (nodeTuple.nodes.hasNext()) {
-                            listener.trackFailure(responseOrResponseException.responseException);
-                            performRequestAsync(nodeTuple, request, listener);
+        request.cancellable.runIfNotCancelled(() -> {
```

Comment:
This provided runnable is executed synchronously so I don't see how the cancellable can already be cancelled ?
In other word, when is this check needed ?

Reply:
the synchronized bit is important here. one can try to cancel when we are in between different attempts of the same request: internally, there is nothing to cancel hence abort will do nothing in that case. I want to make sure that calling cancel with our client gives some guarantee that it will have some effect.